felix/bpf-gpl/tc.c

```diff
@@ -1431,10 +1432,26 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
 				 (CALI_F_TO_HEP &&
 				  ((CALI_F_DSR && skb_seen(ctx->skb)) || !skb_seen(ctx->skb))));
 
-			/* ... then fix the outer header IP first */
 			if (outer_ip_snat) {
 				ip_hdr_set_ip(ctx, saddr, state->ct_result.nat_ip);
-#ifndef IPVER6
+#ifdef IPVER6
+				/* ... icmp6 has checksum now, fix it! */
+				l4_csum_off = skb_l4hdr_offset(ctx) + offsetof(struct icmp6hdr, icmp6_cksum);
+
+				__wsum csum = 0;
+				csum = bpf_csum_diff((__u32*)&STATE->ip_src, sizeof(ipv6_addr_t),
+						     (__u32*)&STATE->ct_result.nat_ip, sizeof(ipv6_addr_t),
+						     csum);
+				csum = bpf_csum_diff((__u32*)&STATE->ip_dst, sizeof(ipv6_addr_t),
+						     (__u32*)&STATE->ct_result.nat_sip, sizeof(ipv6_addr_t),
+						     csum);
+				int res = bpf_l4_csum_replace(ctx->skb, l4_csum_off, 0, csum,  BPF_F_PSEUDO_HDR);
+				if (res) {
+					deny_reason(ctx, CALI_REASON_CSUM_FAIL);
+					goto deny;
+				}
+#else
+				/* ... then fix the outer header IP first */
```

Comment:
first?

Reply:
This is original ipv4 code/comment, we do nat, we need to first fix the outter header before we move to the inner header in the rest of the code. But the comment should move back forward.